src/XMakeBuildEngine/BackEnd/Components/Communications/NodeProviderInProc.cs

```diff
@@ -203,19 +203,19 @@ public bool CreateNode(int nodeId, INodePacketFactory factory, NodeConfiguration
             // Attempt to get the operating environment semaphore if requested.
             if (_componentHost.BuildParameters.SaveOperatingEnvironment)
             {
-                // We can only create additional in-proc nodes if we have decided not to save the operating environment.  This is the global
-                // DTAR case in Visual Studio, but other clients might enable this as well under certain special circumstances.
-                ErrorUtilities.VerifyThrow(_inProcNodeOwningOperatingEnvironment == null, "Unexpected non-null in-proc node semaphore.");
-
                 if (Environment.GetEnvironmentVariable("MSBUILDINPROCENVCHECK") == "1")
                 {
                     _exclusiveOperatingEnvironment = true;
                 }
 
                 if (_exclusiveOperatingEnvironment)
                 {
-                    _inProcNodeOwningOperatingEnvironment = new Semaphore(1, 1, "MSBuildIPN_" + Process.GetCurrentProcess().Id);
-                    if (!_inProcNodeOwningOperatingEnvironment.WaitOne(0))
+                    if (InProcNodeOwningOperatingEnvironment == null)
+                    {
+                        InProcNodeOwningOperatingEnvironment = new Semaphore(1, 1);
+                    }
+
+                    if (!InProcNodeOwningOperatingEnvironment.WaitOne())
```

Comment:
Should this continue to be `WaitOne(0)`? Without a zero timeout, this call will block, which doesn't seem like the right behavior since nothing will release the semaphore until node shutdown.